File: dashboard/src/lib/auditReport.ts
import fs from "fs";
import path from "path";

import type { LeaderboardRow } from "./api";

// Must match arena.core.config.REPORT_SCHEMA_VERSION. The producer embeds this in
// the JSON; a mismatch means the dashboard and the CLI are out of step.
export const EXPECTED_REPORT_SCHEMA_VERSION = "1.0";

export type AuditReviewerRow = {
  reviewer: string;
  model: string;
  mode: "review" | "patch" | "full";
  detection_precision: number | null;
  detection_recall: number | null;
  detection_f_beta: number | null;
  validated_precision: number | null;
  validated_recall: number | null;
  validated_f_beta: number | null;
  validated_case_rate: number | null;
  complete_repair_rate: number | null;
  bug_completeness_rate: number | null;
  supported_claim_rate: number | null;
  deterministic_pass_rate: number | null;
  patch_apply_rate: number | null;
  test_pass_rate: number | null;
  structural_pass_rate: number | null;
  false_positives_per_case: number | null;
  cost_per_validated_fix: number | null;
  latency_per_case_ms: number | null;
  run_id: string;
  primary_failure_mode?: string | null;
};

export type AuditReport = {
  schema_version: string;
  generated_at: string;
  empty: boolean;
  summary: {
    benchmark_pack: string;
    run_count: number;
    case_count: number;
    reviewers_tested: string[];
    biggest_detection_validation_gap: { reviewer: string; gap: number } | null;
  };
  reviewers: AuditReviewerRow[];
  failure_modes: Record<string, number>;
  case_studies: Array<{
    case_id: string;
    reviewer: string;
    model: string;
    finding_summary: string;
    failure_reasons: string[];
    validator_evidence: Array<{
      name: string;
      passed: boolean;
      message: string;
    }>;
    test_stderr_tail: string;
  }>;
  reproducibility_commands: string[];
  limitations: string[];
};

export type AuditReportResult = {
  report: AuditReport | null;
  error: string | null;
};

const REGENERATE_HINT = "Regenerate it with `arena audit-report runs/`.";

/**
 * Read an audit report JSON (default audit-v1.json), distinguishing "not generated
 * yet" (report and error both null) from a real failure (error set). Callers that
 * must not show stale or fake data render the error; fallback callers can ignore it
 * and treat null as empty.
 */
export function readAuditReport(fileName = "audit-v1.json"): AuditReportResult {
  const reportFile = path.join(process.cwd(), "public", "reports", fileName);
  if (!fs.existsSync(reportFile)) {
    return { report: null, error: null };
  }
  let parsed: unknown;
  try {
    parsed = JSON.parse(fs.readFileSync(reportFile, "utf-8"));
  } catch (cause) {
    const detail = cause instanceof Error ? cause.message : String(cause);
    return {
      report: null,
      error: `${fileName} is not valid JSON (${detail}). ${REGENERATE_HINT}`,
    };
  }
  const version = (parsed as { schema_version?: unknown }).schema_version;
  if (version !== EXPECTED_REPORT_SCHEMA_VERSION) {
    return {
      report: null,
      error:
        `${fileName} schema_version ${JSON.stringify(version)} does not match the ` +
        `expected ${JSON.stringify(EXPECTED_REPORT_SCHEMA_VERSION)}. ${REGENERATE_HINT}`,
    };
  }
  return { report: parsed as AuditReport, error: null };
}

export function loadReportLeaderboardRows(): LeaderboardRow[] {
  const report = readAuditReport().report;
  return (report?.reviewers ?? [])
    .filter(
      (row) =>
        row.detection_precision != null &&
        row.detection_recall != null &&
        row.validated_precision != null &&
        row.validated_recall != null,
    )
    .map((row) => ({
      reviewer: row.reviewer,
      model: row.model,
      mode: row.mode,
      run_id: row.run_id,
      benchmark_set: "audit_v1",
      detail_available: false,
      score: 0,
      bugs_found: 0,
      case_count: report?.summary.case_count ?? 10,
      false_positives: Math.round(
        (row.false_positives_per_case ?? 0) *
          (report?.summary.case_count ?? 10),
      ),
      cost: 0,
      latency_ms:
        (row.latency_per_case_ms ?? 0) * (report?.summary.case_count ?? 10),
      history_count: 1,
      completed_at: report?.generated_at ?? "",
      deterministic_passes: Math.round(
        (row.deterministic_pass_rate ?? 0) * (report?.summary.case_count ?? 10),
      ),
      deterministic_metrics:
        row.detection_f_beta == null || row.validated_f_beta == null
          ? null
          : {
              detection_precision: row.detection_precision!,
              detection_recall: row.detection_recall!,
              detection_f1: row.detection_f_beta,
              detection_f_beta: row.detection_f_beta,
              validated_precision: row.validated_precision!,
              validated_recall: row.validated_recall!,
              validated_f1: row.validated_f_beta,
              validated_f_beta: row.validated_f_beta,
              beta: 1,
              deterministic_pass_rate: row.deterministic_pass_rate ?? 0,
              validated_case_rate:
                row.validated_case_rate ?? row.deterministic_pass_rate ?? 0,
              complete_repair_rate: row.complete_repair_rate ?? 0,
              bug_completeness_rate: row.bug_completeness_rate ?? 0,
              supported_claim_rate: row.supported_claim_rate,
              patch_apply_rate: row.patch_apply_rate,
              test_pass_rate: row.test_pass_rate,
              structural_pass_rate: row.structural_pass_rate,
              false_positives_per_case: row.false_positives_per_case ?? 0,
              cost_per_validated_fix: row.cost_per_validated_fix,
              latency_per_case_ms: row.latency_per_case_ms ?? 0,
            },
    }));
}

File: arena/reports/audit_report.py
"""Aggregate audit_v1 runs into a shareable Markdown and JSON report."""

from __future__ import annotations

import json
import warnings
from collections import Counter
from datetime import UTC, datetime
from pathlib import Path
from typing import Any

from arena.core.config import REPORT_SCHEMA_VERSION
from arena.core.models import RunResult
from arena.reports.json_report import read_json_report
from arena.reports.report_schema import AuditReport

AUDIT_BENCHMARK_SET = "audit_v1"
GAP_THRESHOLD = 0.15
FAILURE_REASON_LABELS = (
    "patch_required_but_missing",
    "patch_apply_failed",
    "structural_validation_failed",
    "tests_failed",
    "localization_failed",
    "detection_failed",
    "incomplete_bug_detection",
    "false_positive",
)


def load_audit_runs(runs_dir: Path) -> list[RunResult]:
    runs: list[RunResult] = []
    for path in sorted(runs_dir.glob("*/run.json")):
        try:
            run = read_json_report(path)
        except (OSError, ValueError, json.JSONDecodeError) as exc:
            warnings.warn(f"Skipping malformed run file {path}: {exc}", stacklevel=2)
            continue
        if run.benchmark_set == AUDIT_BENCHMARK_SET:
            runs.append(run)
    return runs


def _format_rate(value: float | None) -> str:
    return f"{value * 100:.1f}%" if value is not None else "n/a"


def _reviewer_label(run: RunResult) -> str:
    return f"{run.reviewer}:{run.model}" if run.model else run.reviewer


def build_audit_report_data(runs: list[RunResult]) -> dict[str, Any]:
    if not runs:
        return {
            "schema_version": REPORT_SCHEMA_VERSION,
            "title": "Detection Is Not Validation: Audit Pack v1 Results",
            "generated_at": datetime.now(UTC).isoformat(),
            "empty": True,
            "summary": {
                "benchmark_pack": AUDIT_BENCHMARK_SET,
                "run_count": 0,
                "case_count": 10,
                "reviewers_tested": [],
                "biggest_detection_validation_gap": None,
            },
            "reviewers": [],
            "gaps": [],
            "failure_modes": {},
            "case_studies": [],
            "reproducibility_commands": _reproducibility_commands(),
            "limitations": _limitations(),
        }

    latest: dict[tuple[str, str | None, str], RunResult] = {}
    for run in runs:
        key = (run.reviewer, run.model, run.mode)
        previous = latest.get(key)
        if previous is None or run.completed_at > previous.completed_at:
            latest[key] = run

    reviewer_rows: list[dict[str, Any]] = []
    gaps: list[dict[str, Any]] = []
    failure_counter: Counter[str] = Counter()
    biggest_gap = 0.0
    biggest_gap_label: str | None = None

    for run in latest.values():
        metrics = run.deterministic_metrics
        detection = metrics.detection_f_beta if metrics else None
        validated = metrics.validated_f_beta if metrics else None
        gap = (detection or 0.0) - (validated or 0.0)
        run_failure_counter: Counter[str] = Counter()
        if gap > biggest_gap:
            biggest_gap = gap
            biggest_gap_label = _reviewer_label(run)
        if detection is not None and validated is not None and gap >= GAP_THRESHOLD:
            gaps.append(
                {
                    "reviewer": run.reviewer,
                    "model": run.model or "",
                    "mode": run.mode,
                    "detection_f_beta": detection,
                    "validated_f_beta": validated,
                    "gap": round(gap, 6),
                    "run_id": run.run_id,
                }
            )
        for case in run.case_results:
            for reason in case.failure_reasons:
                if reason in FAILURE_REASON_LABELS:
                    failure_counter[reason] += 1
                    run_failure_counter[reason] += 1
                elif reason == "detection_failed":
                    failure_counter["detection_failed"] += 1
                    run_failure_counter["detection_failed"] += 1
                elif reason == "localization_failed":
                    failure_counter["localization_failed"] += 1
                    run_failure_counter["localization_failed"] += 1
        reviewer_rows.append(
            {
                "reviewer": run.reviewer,
                "model": run.model or "",
                "mode": run.mode,
                "detection_precision": metrics.detection_precision if metrics else None,
                "detection_recall": metrics.detection_recall if metrics else None,
                "detection_f_beta": detection,
                "validated_precision": metrics.validated_precision if metrics else None,
                "validated_recall": metrics.validated_recall if metrics else None,
                "validated_f_beta": validated,
                "deterministic_pass_rate": metrics.deterministic_pass_rate if metrics else None,
                "patch_apply_rate": metrics.patch_apply_rate if metrics else None,
                "test_pass_rate": metrics.test_pass_rate if metrics else None,
                "structural_pass_rate": metrics.structural_pass_rate if metrics else None,
                "false_positives_per_case": metrics.false_positives_per_case if metrics else None,
                "cost_per_validated_fix": metrics.cost_per_validated_fix if metrics else None,
                "latency_per_case_ms": metrics.latency_per_case_ms if metrics else None,
                "run_id": run.run_id,
                "primary_failure_mode": (
                    run_failure_counter.most_common(1)[0][0] if run_failure_counter else None
                ),
            }
        )

    def _validated_sort_key(row: dict[str, Any]) -> float:
        value = row.get("validated_f_beta")
        return float(value) if isinstance(value, (int, float)) else -1.0

    reviewer_rows.sort(key=_validated_sort_key, reverse=True)
    gaps.sort(key=lambda row: row["gap"], reverse=True)
    case_studies = _select_case_studies(list(latest.values()))

    return {
        "schema_version": REPORT_SCHEMA_VERSION,
        "title": "Detection Is Not Validation: Audit Pack v1 Results",
        "generated_at": datetime.now(UTC).isoformat(),
        "empty": False,
        "summary": {
            "benchmark_pack": AUDIT_BENCHMARK_SET,
            "run_count": len(runs),
            "case_count": max((run.case_count for run in latest.values()), default=10),
            "reviewers_tested": sorted({_reviewer_label(run) for run in latest.values()}),
            "biggest_detection_validation_gap": (
                {"reviewer": biggest_gap_label, "gap": round(biggest_gap, 6)}
                if biggest_gap_label
                else None
            ),
        },
        "reviewers": reviewer_rows,
        "gaps": gaps,
        "failure_modes": dict(failure_counter),
        "case_studies": case_studies,
        "reproducibility_commands": _reproducibility_commands(),
        "limitations": _limitations(),
    }


def _select_case_studies(runs: list[RunResult]) -> list[dict[str, Any]]:
    candidates: list[tuple[int, dict[str, Any]]] = []
    priority = {
        "structural_validation_failed": 4,
        "tests_failed": 3,
        "patch_apply_failed": 2,
        "patch_required_but_missing": 2,
        "localization_failed": 1,
        "detection_failed": 1,
        "false_positive": 1,
    }
    for run in runs:
        for case in run.case_results:
            if case.deterministic_pass is True:
                continue
            reasons = case.failure_reasons
            if not reasons:
                continue
            top_reason = max(reasons, key=lambda reason: priority.get(reason, 0))
            detection_only = case.bug_found and case.correct_file and case.correct_line
            validated = case.deterministic_pass is True
            if detection_only and not validated:
                top_reason = reasons[0]
            finding_summary = ""
            if case.scored_findings:
                finding_summary = case.scored_findings[0].finding.summary
            candidates.append(
                (
                    priority.get(top_reason, 0),
                    {
                        "case_id": case.case_id,
                        "reviewer": run.reviewer,
                        "model": run.model or "",
                        "finding_summary": finding_summary,
                        "failure_reasons": reasons,
                        "validator_evidence": [
                            {
                                "name": item.get("name"),
                                "passed": item.get("passed"),
                                "message": item.get("message"),
                            }
                            for item in case.validator_results[:2]
                        ],
                        "test_stderr_tail": case.test_stderr_tail[-400:]
                        if case.test_stderr_tail
                        else "",
                    },
                )
            )
    candidates.sort(key=lambda item: item[0], reverse=True)
    return [item[1] for item in candidates[:3]]


def _reproducibility_commands() -> list[str]:
    return [
        "arena validate benchmark_sets/audit_v1",
        "arena run benchmark_sets/audit_v1 --reviewer reference-patch --mode full --allow-local-execution",
        "arena run benchmark_sets/audit_v1 --reviewer control:keyword_gamer --mode full --allow-local-execution",
        "arena leaderboard runs/ --metric validated_f_beta --beta 1.0",
        "arena audit-report runs/ --output docs/reports/audit-v1-results.md",
    ]


def _limitations() -> list[str]:
    return [
        "audit_v1 is curated and small; it measures a narrow set of patch-backed failures.",
        "Structural validators are hand-authored and may reject alternate but valid repairs.",
        "Regression tests cannot prove full production correctness.",
        "Some credible fixes may fail when validators are intentionally narrow.",
        "This pack is not designed to compete with large public leaderboard scale benchmarks.",
    ]


def render_audit_report_markdown(data: dict[str, Any]) -> str:
    lines = [f"# {data['title']}", ""]
    if data.get("empty"):
        lines.extend(
            [
                "_No audit_v1 runs were found under the requested runs directory._",
                "",
                "## Reproducibility",
                "",
            ]
        )
        lines.extend(f"- `{command}`" for command in data["reproducibility_commands"])
        lines.extend(["", "## Limitations", ""])
        lines.extend(f"- {item}" for item in data["limitations"])
        return "\n".join(lines) + "\n"

    summary = data["summary"]
    lines.extend(
        [
            "## 1. Summary",
            "",
            f"- Benchmark pack: `{summary['benchmark_pack']}`",
            f"- Runs discovered: {summary['run_count']}",
            f"- Cases per run: {summary['case_count']}",
            f"- Reviewers tested: {', '.join(summary['reviewers_tested']) or 'none'}",
        ]
    )
    gap = summary.get("biggest_detection_validation_gap")
    if gap:
        lines.append(
            f"- Biggest detection-validation gap: `{gap['reviewer']}` "
            f"(detection_f_beta - validated_f_beta = {gap['gap']:.3f})"
        )
    lines.extend(["", "## 2. Methodology", ""])
    lines.extend(
        [
            "- Ground truth stays hidden from reviewer prompts and custom-command JSON.",
            "- Reviewers receive the PR diff, relevant files, and optional test/static output.",
            "- Full mode applies `suggested_patch`, runs regression tests, and structural validators.",
            "- `detection_f_beta` scores localization only.",
            "- `validated_f_beta` is the primary full/patch metric and requires deterministic pass.",
            "",
            "## 3. Reviewer Comparison",
            "",
            "| Reviewer | Model | Mode | Detection F-beta | Validated F-beta | "
            "Deterministic Pass Rate | Patch Apply Rate | Test Pass Rate | "
            "Structural Pass Rate | False Positives / Case | Cost / Validated Fix | Latency / Case |",
            "|---|---|---|---:|---:|---:|---:|---:|---:|---:|---:|---:|",
        ]
    )
    for row in data["reviewers"]:
        lines.append(
            f"| {row['reviewer']} | {row['model']} | {row['mode']} | "
            f"{_metric_cell(row.get('detection_f_beta'))} | {_metric_cell(row.get('validated_f_beta'))} | "
            f"{_format_rate(row.get('deterministic_pass_rate'))} | "
            f"{_format_rate(row.get('patch_apply_rate'))} | "
            f"{_format_rate(row.get('test_pass_rate'))} | "
            f"{_format_rate(row.get('structural_pass_rate'))} | "
            f"{row.get('false_positives_per_case', 'n/a')} | "
            f"{row.get('cost_per_validated_fix', 'n/a')} | "
            f"{row.get('latency_per_case_ms', 'n/a')} |"
        )
    lines.extend(["", "## 4. Detection vs Validation Gap", ""])
    if data["gaps"]:
        for gap in data["gaps"]:
            lines.append(
                f"- `{gap['reviewer']}:{gap['model']}` ({gap['mode']}): detection={gap['detection_f_beta']:.3f}, "
                f"validated={gap['validated_f_beta']:.3f}, gap={gap['gap']:.3f} "
                f"(run `{gap['run_id']}`)"
            )
    else:
        lines.append("- No runs exceeded the configured detection-validation gap threshold.")
    lines.extend(["", "## 5. Failure Mode Breakdown", ""])
    if data["failure_modes"]:
        for reason, count in sorted(data["failure_modes"].items(), key=lambda item: -item[1]):
            lines.append(f"- `{reason}`: {count}")
    else:
        lines.append("- No recorded failure reasons.")
    lines.extend(["", "## 6. Case Studies", ""])
    if data["case_studies"]:
        for study in data["case_studies"]:
            lines.append(f"### {study['case_id']} ({study['reviewer']}:{study['model']})")
            if study.get("finding_summary"):
                lines.append(f"- Finding: {study['finding_summary']}")
            lines.append(f"- Failure reasons: {', '.join(study['failure_reasons'])}")
            if study.get("validator_evidence"):
                lines.append("- Validator evidence:")
                for item in study["validator_evidence"]:
                    lines.append(f"  - `{item['name']}`: {item['message']}")
            if study.get("test_stderr_tail"):
                lines.append(f"- Test output tail: `{study['test_stderr_tail']}`")
            lines.append("")
    else:
        lines.append("- No failing case studies were available.")
    lines.extend(["", "## 7. Reproducibility", ""])
    lines.extend(f"- `{command}`" for command in data["reproducibility_commands"])
    lines.extend(["", "## 8. Limitations", ""])
    lines.extend(f"- {item}" for item in data["limitations"])
    return "\n".join(lines) + "\n"


def _metric_cell(value: float | None) -> str:
    return f"{value:.3f}" if value is not None else "n/a"


def write_audit_report(
    runs_dir: Path,
    output_path: Path,
    json_output_path: Path | None = None,
) -> dict[str, Any]:
    data = build_audit_report_data(load_audit_runs(runs_dir))
    # Validate the contract before writing so a producer-side drift fails loudly here
    # rather than silently in the dashboard.
    AuditReport.model_validate(data)
    output_path.parent.mkdir(parents=True, exist_ok=True)
    output_path.write_text(render_audit_report_markdown(data), encoding="utf-8")
    if json_output_path is not None:
        json_output_path.parent.mkdir(parents=True, exist_ok=True)
        json_output_path.write_text(json.dumps(data, indent=2), encoding="utf-8")
    return data
